images/airflow/2.9.2/python/mwaa/celery/task_monitor.py

```diff
@@ -376,8 +518,17 @@ def pause_task_consumption(self):
             )
             return
 
-        logger.info("Pausing task consumption.")
+        was_consumption_unpaused = self.celery_work_consumption_block.buf[0] == 0
         self.celery_work_consumption_block.buf[0] = 1
+        if was_consumption_unpaused:
+            # When we toggle the consumption to paused state, we wait 10 seconds in order
+            # for any in-flight messages in the SQS broker layer to be processed and
+            # corresponding Airflow task instance to be created. Once that is done, we can
+            # start gracefully shutting down the worker. Without this, the SQS broker may
+            # consume messages from the queue, terminate before creating the corresponding
+            # Airflow task instance and abandon SQS messages in-flight.
+            logger.info("Pausing task consumption.")
+            time.sleep(10)
```

Comment:
Why 10 seconds?

Reply:
I was thinking of keeping a 10 second value for the case where we resume the consumption. My thought process was that these 10 seconds will allow enough time for the worker to pick any work if available from the queue. This will cause the worker to become busy after its consumption has been resumed. And from there, I thought of keeping things consistent between the resume and the pause cases.

But thinking about this again, we have been using 5 seconds for the pause case with improved autoscaling with no issues. Perhaps, I should change them both to 5 seconds.